### PR TITLE
Refactor application helper

### LINKED ISSUE #219: Refactor Application Helper and add Unit Tests
This helper currently has a few long methods which could use some refactoring. This also needs a unit test.

https://github.com/TheOdinProject/theodinproject/blob/master/app/helpers/application_helper.rb

### PR DESCRIPTION
**Pull request by arun1595**
Fixes #219

### CODE REVIEW

File: spec/helpers/application_helper_spec.rb [hunk 1]
```diff
@@ -25,4 +25,27 @@
...
+    end
+
+    context "non-nil input" do
+      it 'returns nil for an empty string' do
+        expect(title('')).to be nil
+      end
+
+      it 'returns the title for a non-empty string' do
+        helper.title('Courses')
+        expect(helper.instance_variable_get(:@view_flow).
```
Comment (KevinMulhern) [inline spec/helpers/application_helper_spec.rb]: This will leave this test quite brittle, if the name of this variable is ever changed this test will break.

I did a bit of research on testing content_for just now. I don't really like using content for in the tests, but this is probably the cleanest way to test it without stubbing out content_for.

```
it 'returns correctly formatted title' do
      helper.title('hello')
      expect(helper.content_for(:title)).to eql('hello | The Odin Project')
    end
```

File: spec/helpers/application_helper_spec.rb [hunk 2]
```diff
@@ -25,4 +25,27 @@
...
+
+  describe "title" do
+
+    context "nil input" do
+      it 'returns nil' do
+        expect(title).to be nil
+      end
+    end
+
+    context "non-nil input" do
```
Comment (KevinMulhern) [inline spec/helpers/application_helper_spec.rb]: for context descriptions its conventional to use the format "**when** input is nil" etc

File: spec/helpers/application_helper_spec.rb [hunk 3]
```diff
@@ -25,4 +25,27 @@
       end
     end
   end
+
+  describe "title" do
+
+    context "nil input" do
```
Comment (KevinMulhern) [inline spec/helpers/application_helper_spec.rb]: Can we move this to be the last test in the describe block please?. The happy path should be first ie test that it works with an input and then test the other branches and edge cases.

File: app/views/courses/index.html.erb [hunk 4]
```diff
@@ -1,4 +1,5 @@
-<%= content_for :title, title("Free Web Development Courses") %>
+<%= title("Courses") %>
```
Comment (KevinMulhern) [inline app/views/courses/index.html.erb]: can we have single quotes instead please

File: app/helpers/application_helper.rb [hunk 5]
```diff
@@ -5,22 +5,9 @@ def chat_link
...
-      title += "Learn #{@course.title}"
-      if @lesson
-        title += " -- #{@lesson.title}"
-      else
-        title += " for Free"
-      end
-    else
-      title = "Learn Web Development for Free Using Ruby on Rails"
-      title += "  |  The Odin Project"
+    if input and !input.empty?
```
Comment (arun1595) [inline app/helpers/application_helper.rb]: Not really. I was just being defensive. Should I remove the empty check?

File: spec/helpers/application_helper_spec.rb (hunk 1 above)
Comment (arun1595) [inline spec/helpers/application_helper_spec.rb]: > if the name of this variable is ever changed this test will break.

I don't really understand what you mean.

File: app/helpers/application_helper.rb (hunk 5 above)
Comment (KevinMulhern) [inline app/helpers/application_helper.rb]: Yeah I think its safe to remove.

File: spec/helpers/application_helper_spec.rb (hunk 1 above)
Comment (KevinMulhern) [inline spec/helpers/application_helper_spec.rb]: Oh sorry, I thought I edited that part. if `:@view_flow` ever changes in a future rails version is what I meant.

File: app/views/courses/index.html.erb (hunk 4 above)
Comment (arun1595) [inline app/views/courses/index.html.erb]: Done.

File: spec/helpers/application_helper_spec.rb (hunk 1 above)
Comment (arun1595) [inline spec/helpers/application_helper_spec.rb]: Oh, I understood you right then :) Fine, I have modified the tests.

**Comment by KevinMulhern** [comment]
👍 LGTM
